test/src/utils/quaternion.test.cpp

```diff
@@ -0,0 +1,353 @@
+{
+  float sumSquare = 0;
+  for (int i = 0; i < 4; i++) {
+    sumSquare += pow(quat[i], 2);
+  }
+  return sqrt(sumSquare);
+}
+
+// -------------------------------------------------------------------------------------------------
+// Functionality
+// -------------------------------------------------------------------------------------------------
+
+/**
+ * Struct used for testing the functionality
+ * of quaternion.hpp
+ */
+struct QuaternionFunctionality : public::testing::Test
+{
+    /* data */
+    const int kSizeOfQuaternion = 4;
+    Quaternion<int> quaternion_one;
+    Quaternion<int> quaternion_two;
+
+    Quaternion<int> quaternion_result_one;
+    Quaternion<int> quaternion_result_two;
+
+    Vector<int, 4> test_vector;
+    Vector<int, 3> test_vector_3d;
+
+    float arithmetic_result;
+
+    int kValue;
+
+    void SetUp()
+    {
+        kValue = rand()%1000;
+        quaternion_one = Quaternion<int>(rand()%1000, rand()%1000, rand()%1000, rand()%1000);
+        quaternion_two = Quaternion<int>(rand()%1000, rand()%1000, rand()%1000, rand()%1000);
+    }
+};
+
+// -------------------------------------------------------------------------------------------------
+// Tests
+// -------------------------------------------------------------------------------------------------
+
+/**
+ * @brief Test to determine the correct instantiation of the zero
+ * Quaternion class instance
+ */
+TEST_F(QuaternionFunctionality, handlesConstructionOfZeroQuaternion)
+{
+    Quaternion<int> zeroQuaternion = Quaternion<int>(0);
+    ASSERT_EQ(0, zeroQuaternion.norm());
+    for (int i = 0; i < kSizeOfQuaternion; i++) {
+        ASSERT_EQ(0, zeroQuaternion[i]);
+    }
+}
+
+/**
+ * @brief Test to determine the correct instantiation of a non zero
+ * Quaternion class instance.
+ */
+TEST_F(QuaternionFunctionality, handlesConstructionOfNonZeroQuaternion)
+{
+  int component_array[kSizeOfQuaternion];
+  for (int i = 0; i < kSizeOfQuaternion; i++) {
+    component_array[i] = rand()%1000;
+    while (component_array[i] == 0) {
+      component_array[i] = rand()%1000;
+    }
+  }
+  Quaternion<int> nonZeroQuaternion = Quaternion<int>(
+    component_array[0],
+    component_array[1],
+    component_array[2],
+    component_array[3]);
+  for (int i = 0; i < kSizeOfQuaternion; i++) {
+    ASSERT_EQ(nonZeroQuaternion[i], component_array[i]);
+  }
+}
+
+/**
+ * @brief Test to determine the correct construction of Quaternion
+ * instances by Vector instances. This test includes contruction by
+ * a 4D Vector, 3D Vector and constant, and 3D Vector.
+ */
+TEST_F(QuaternionFunctionality, handlesConstructionByVector)
+{
+  // Testing construction by 4D Vector
+  test_vector = Vector<int, 4>(createRandomArrayForQuaternion());
+  Quaternion<int> quaternion_vector = Quaternion<int>(test_vector);
+  for (int i = 0; i < kSizeOfQuaternion; i++) {
+    ASSERT_EQ(test_vector[i], quaternion_vector[i]);
+  }
+
+  std::array<int, 3> component_array;
```

Comment:
Maybe this can be extracted into a helper function

Reply:
Agreed. I tried to modify the `createRandomArrayForQuaternion` function to vary the size of the array it returned based on an argument, but the size of the array returned is defined in the type returned.